ncclient/logging_.py

```diff
@@ -0,0 +1,26 @@
+import logging
+
+class SessionLoggerAdapter(logging.LoggerAdapter):
+    """Logger adapter that automatically adds session information to logs."""
+
+    def process(self, msg, kwargs):
```

Comment:
Will this function be evaluated every time we call `self.logger.debug`, etc., or will it only be called if the debug level used in the client code is enabled?

Reply:
Fair question. I don't know but I can look into it. Concerned about performance? All I really *need* for my purposes is the inclusion of 'session' in the logrecord `extra` data; if you want to avoid the string processing overhead I can remove that part.

Reply:
https://github.com/python/cpython/blob/master/Lib/logging/__init__.py#L1721-L1728:

```python
    def log(self, level, msg, *args, **kwargs):
        """
        Delegate a log call to the underlying logger, after adding
        contextual information from this adapter instance.
        """
        if self.isEnabledFor(level):
            msg, kwargs = self.process(msg, kwargs)
            self.logger.log(level, msg, *args, **kwargs)
```

So `process` is only called if the relevant logging level is enabled.

Reply:
Yes, was thinking about performance. Now I'm happy :-) Can you resolve the conficts now introduced with master branch? Then it'll be good to go!